### ISSUE TITLE
Improve Singletons

### ISSUE THREAD

**Issue by martingrumbt** [k3b/APhotoManager#182]
**Describe the bug**

I analysed your project for design patterns.
The implementation of the Singleton pattern in some classes is not done in the common way.
The following classes can be improved:

- TagRepository
- PhotoPropertiesMediaFilesScanner
- DatabaseHelper

**Expected behavior**

- Avoid public constructor
- Only allow access via getInstance-Method
- Use synced variant to be thread safe
- Avoid Setter to change Instance

**Additional context**

Here are some blueprints for the Implementation that are all thread safe:

_Lazy Instantiation_

```java
public class LazySingleton {
    private static volatile LazySingleton instance;

    private LazySingleton(args) { /* ... */ }

   
    public static LazySingleton getInstance(args) {
        if (instance == null) {
            synchronized (LazySingleton.class) {
                if (instance == null) {
                    instance = new LazySingleton(args);
                }
            }
        }
        return instance;
    }

     /* other methods */
}
```

Pros:
- arguments can be forwarded on instantiation

_Enum Instantiation_

```java
public enum EnumSingleton {
    INSTANCE(args); 
 
    private EnumSingleton(args) {
        /* ... */ 
    }
 
    public EnumSingleton getInstance() {
        return INSTANCE;
    }
    
    /* other methods */
}
```

Pros:
- best way when no arguments are forwarded on instantiation

Cons:
- no forwarding of arguments to the constructor when used first time -> Setters needed

_Early Instantiation_

```java
public class EarlySingleton {
    public static final EarlySingleton INSTANCE = new EarlySingleton(args); 

    private EarlySingleton(args) { /* ... */ }

    /* other methods */
}
```

Cons:
- instantiation when class is loaded
- no forwarding of arguments to the constructor when used first time -> Setters needed

**Comment by k3b** [comment]
Thanks for taking the time to analyse my code.

> The implementation of the Singleton pattern in some classes is not done in the common way.
> ...
> Expected behavior
> ...
> Avoid Setter to change Instance

Analysis:

* DatabaseHelper follows your pattern of "Lazy Instantiation" without "synchronized " with passing needed parameters. 
   * the user of the singelton knows about parameter needed to create the sigleton.
* TagRepository, PhotoPropertiesMediaFilesScanner follow a variation of your pattern "Early Instantiation" that is not possible with your 3 example implementation.
   * forwarding of arguments to the constructor needed to create the sigleton and 
  * the Singletons should have no dependency to global configuration and
  * the user of the singelton should not know anything about global configuration.
  * all configuration specific singleton-initialisatztion (and other inititalisations) are done in one class outside singleton and outside "global configuration". This is "Early Instantiation in a seperate class"
    * in java jee applications this is usually done through Dependency-Injection-Frameworks (i.e. spring)
    * to keep the memory foodprint of my android app small i donot use Dependency-Injection-Frameworks so i do the wiring-up on my own.

You should update your variations patterns:

* Lazy Instantiation - cons: the user of the singelton must know parameters (configuration data) needed  to initializee
* rename "Early Instantiation" to "Early Instantiation with dependency to configuration data"

**Comment by k3b** [comment]
findings from reviewing this issue:

move code for "Lazy Instantiation" from SettingsActivity to seperate class (Architectural flaw)

**Comment by martingrumbt** [comment]
It was not as successful as I hoped the requests would be.

Github flagged my account because I opened too many issues with similar content. Most of the issues were simply closed or it was requested that the changes be submitted as a pull request.

--- 
Now to your question:
Corresponding to the topic for my bachelor thesis I optimised the search for the singleton pattern of the JTransformer pattern recognition in Eclipse.

The results I wanted to share with the projects I analysed.
The analyses is not broad enough to involve the context of each Singleton. 
So it's hard to decide if a varying implementation of the pattern leads to unwanted side effects or the developer decided it conscious.

That was the point to open the issues. So each project maintainer can have a look and decide to fix it or not.

Thank you for responding.
